plugins/out_stackdriver/stackdriver_source_location.c

```diff
@@ -0,0 +1,158 @@
+            || strncmp(SOURCELOCATION_FIELD_IN_JSON, 
+                       p->key.via.str.ptr, p->key.via.str.size) != 0) {
+
+            continue;
+        }
+
+        op_status = SOURCELOCATION_EXISTED;
+        msgpack_object sub_field = p->val;
+        
+        tmp_p = sub_field.via.map.ptr;
+        tmp_pend = sub_field.via.map.ptr + sub_field.via.map.size;
+
+        /* Validate the subfields of sourceLocation */
+        for (; tmp_p < tmp_pend; ++tmp_p) {
+            if (tmp_p->key.type != MSGPACK_OBJECT_STR) {
+                continue;
+            }
+
+            if (sizeof("file") - 1 == tmp_p->key.via.str.size
+                && strncmp("file", tmp_p->key.via.str.ptr, 
+                           tmp_p->key.via.str.size) == 0) {
+
+                if (tmp_p->val.type != MSGPACK_OBJECT_STR) {
+                    continue;
+                }
+                *source_location_file = flb_sds_copy(*source_location_file, 
+                                                     tmp_p->val.via.str.ptr, 
+                                                     tmp_p->val.via.str.size);
+            }
+            else if (sizeof("function") - 1 == tmp_p->key.via.str.size
+                     && strncmp("function", tmp_p->key.via.str.ptr, 
+                                tmp_p->key.via.str.size) == 0) {
+
+                if (tmp_p->val.type != MSGPACK_OBJECT_STR) {
+                    continue;
+                }
+                *source_location_function = flb_sds_copy(*source_location_function, 
+                                                         tmp_p->val.via.str.ptr, 
+                                                         tmp_p->val.via.str.size);
+            }
+            else if (sizeof("line") - 1 == tmp_p->key.via.str.size
+                     && strncmp("line", tmp_p->key.via.str.ptr, 
+                                tmp_p->key.via.str.size) == 0) {
+
+                if (tmp_p->val.type == MSGPACK_OBJECT_STR) {
+                    *source_location_line = atoll(tmp_p->val.via.str.ptr);
+                }
+                else if (tmp_p->val.type == MSGPACK_OBJECT_POSITIVE_INTEGER) {
+                    *source_location_line = tmp_p->val.via.i64;
+                }
+            }
+            else {
+                *extra_subfields += 1;
+            }
+
+        }
+    }
+    
+    
+    return op_status == SOURCELOCATION_EXISTED;
+}
+
+void pack_extra_source_location_subfields(msgpack_packer *mp_pck, 
+                                          msgpack_object *source_location, 
+                                          int extra_subfields) {
+    msgpack_object_kv *p = source_location->via.map.ptr;
+    msgpack_object_kv *const pend = source_location->via.map.ptr + source_location->via.map.size;
+
+    msgpack_pack_map(mp_pck, extra_subfields);
+
+    for (; p < pend; ++p) {
+        if ((sizeof("file") - 1 == p->key.via.str.size
```

Comment:
```suggestion
        if ((sizeof("file") - 1 != p->key.via.str.size
```

Should apply to the 3 conditions below too.

Maybe you can check whether the key is one of them file/line/function - and if it is - `continue` so you don't pack this object? In this way, the logic is clear and easy to understand.